standards/stef_and_luka-standard_proposal_draft.md

```diff
@@ -0,0 +1,274 @@
+
+1. Each subscription ID is calculated off-chain by hashing subscription data: `hash(providerAddress, userAddress, externalSubsID, amount, etc.)`. This is then supplied as part of the create subscription transaction. There are two drawbacks to this:
+    - Calculating globally unique IDs is left to the provider. This is not reliable and rouge calculations (leaving out provider addres, user address and others) can result in on-chain collisions, which severly degrade user experience and require additional off-chain logic to handle them.
+    - Front-running DoS attack. A malicious entity can deprive an honest provider from subscribing new users. Each time a new "create subscription" transaction is broadcast to the network, malicious entity front-runs the transaction and creates a subscription with the same ID with no means of actually paying (e.g. ERC20 allowance set to 0, or setting it to 0 shortly after). The result is that all of the newly created subscriptions fail since they're duplicated by a malicious entity.
+    - To solve both, there has to be on-chain code which re-calculates the hash and confirms subscription ID is calculcated with the right parameters and it belongs to the sender. However, it raises a question of why doing it off-chain if on-chain re-calculation has to happen anyway.
+1. Each subscription ID is supplied by provider and standard treats them as black-boxes.
+    - This is the most loose approach and prone to on-chain collisions. It has the same drawbacks as 1.
+1. Each subscription ID is calculated on-chain from a tuple `(providerAddress, externalId)`. External ID can be provider's external subscription ID or hash of subscription parameters or something else entirely.
+    - Moving everything to be done on-chain prevents front-running attacks. We pass subscription parameters to the "create subscription" function and globally unique ID is generated on-chain. Using at least (`providerAddress`, `externalSubscriptionID`) also guarantees there are no ID collisions (NOTE: we rely on the fact that `externalSubscriptionID` supplied by provider is unique).
+
+
+### ERC165 Interface
+
+Standard Interface Detection (ERC165) is used to expose the interfaces that a ERC948 smart contract supports.
+
+### GAS cost and consumption
+
+This proposal supports implementations which can manage arbitrary large number of subscriptions. It avoids defining functions which would require implementations to use for/while loops in the code (e.g. returning all user subscription IDs, etc.). These indicate implementations may not be able to scale and gas costs will rise indefinitely.
+
+
+## ERC948 Interface
+
+```solidity
+  interface ERC948 /* is ERC165 */ {
+
+  /**
+   * @notice User is determined by msg.sender.
+   * @dev Return user subscription providers if there are any.
+   * @param _user User's address.
+   * @return Array of provider addresses.
+   **/
+  function getUserSubscriptionProviders(address _user) external view
```

Comment:
What's actually the definition for the provider? Is that the destination address for the payment?

Reply:
Yes. I agree. A broader term like "destination address" would be better.

Reply:
Yes, we identify providers by their destination address.